src/sparsezoo/analysis/analysis.py

```diff
@@ -0,0 +1,439 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List, Optional, Union
+
+import onnx
+import yaml
+from onnx import ModelProto, NodeProto
+from pydantic import BaseModel, Field
+
+from sparsezoo.utils import (
+    NodeDataType,
+    NodeShape,
+    extract_node_id,
+    extract_node_shapes_and_dtypes,
+    get_layer_and_op_counts,
+    get_node_four_block_sparsity,
+    get_node_num_four_block_zeros_and_size,
+    get_node_num_zeros_and_size,
+    get_node_sparsity,
+    get_node_weight,
+    get_node_weight_name,
+    get_node_weight_shape,
+    get_num_dense_and_sparse_ops,
+    get_zero_point,
+    is_parameterized_prunable_layer,
+    is_quantized_layer,
+    is_sparse_layer,
+)
+
+
+__all__ = [
+    "NodeAnalysis",
+    "ModelAnalysis",
+]
+
+
+class NodeIO(BaseModel):
+    """
+    Pydantic model for the inputs and outputs of a node in the onnx model graph
+    """
+
+    name: str = Field(description="Name of this input/output in onnx model graph")
+    shape: Optional[List[Union[None, int]]] = Field(
+        description="Shape of this input/output in onnx model graph (assuming a "
+        "batch size of 1)"
+    )
+    dtype: Optional[str] = Field(
```

Comment:
YMMV and certainly everyone won't agree with me, but I'd sooner call this `data_type` than `dtype`